Dockerfile

```diff
@@ -10,16 +10,33 @@ LABEL name="custodian" \
 # to minimize size
 ADD setup.py README.md requirements.txt /src/
 ADD c7n /src/c7n/
-ADD tools /src/tools/
+ADD tools/c7n_gcp /src/tools/c7n_gcp
+ADD tools/c7n_azure /src/tools/c7n_azure
+ADD tools/c7n_mailer /src/tools/c7n_mailer
 
 WORKDIR /src
 
 RUN adduser --disabled-login custodian
-RUN apt-get --yes update && apt-get --yes upgrade \
- && apt-get --yes install build-essential \
+RUN apt-get --yes update \
+ && apt-get --yes install build-essential --no-install-recommends \
  && pip3 install -r requirements.txt  . \
  && pip3 install -r tools/c7n_gcp/requirements.txt tools/c7n_gcp \
  && pip3 install -r tools/c7n_azure/requirements.txt tools/c7n_azure \
+ && pip3 install -r tools/c7n_mailer/requirements.txt tools/c7n_mailer \
+ # Pre-cache Azure Functions package
+ && python -c "from c7n_azure.function_package import FunctionPackage; \
+      FunctionPackage('cache').build_cache( \
+      modules=['c7n', 'c7n-azure', 'applicationinsights'], \
```

Comment:
the actual work thats being done here feels like it could be simplified a bit for the use case here. we already a local directory with all the downloaded files (pip's download cache) for the right os/arch (well presumably arch), it just needs assembling into a zip (install and tree walk), but even that could be simplified here.. ie literally the function cache is basically just a zip of site-packages  sans the `.c` and `.pyc` afaics.

Reply:
> it could be simplified a bit for the use case here.

It could be for sure.  We used to do it that way.  This way is just more robust with os/arch and ensuring clean dependencies with no extra files - none of which is required in a perfect docker image scenario.  But building a working package and testing it is challenging so I don't think we want to support two implementations.

Of course since I went overboard with the mailer package half of this goes away.

Reply:
on the flip side .. its cost of implementation.. its two lines of shell.. that avoids a lot of custom logic that's irrelevant here.